pkg/rm/tcc/fence/handler/tcc_fence_dbproxy_handler.go

```diff
@@ -0,0 +1,213 @@
+	branchId := tm.GetBusinessActionContext(ctx).BranchId
+	actionName := tm.GetBusinessActionContext(ctx).ActionName
+
+	defer func() {
+		rec := recover()
+		err, ok := rec.(seataErrors.TccFenceError)
+		if ok {
+			if err.Code == seataErrors.FenceErrorCodeDuplicateKey {
+				handler.AddToLogCleanQueue(xid, branchId)
+			}
+		}
+		// panic uniform process in outside method
+		panic(rec)
+	}()
+
+	ok := handler.InsertTCCFenceLog(tx, xid, branchId, actionName, constant.StatusTried)
+	log.Infof("tcc fence prepare result: %b. xid: %s, branchId: %d", ok, xid, branchId)
+	if ok {
+		err := callback()
+		if err != nil {
+			if rollbackErr := tx.Rollback(); rollbackErr != nil {
+				return errors.New(err.Error() + rollbackErr.Error())
+			}
+			return err
+		}
+		return tx.Commit()
+	} else {
+		return seataErrors.NewTccFenceError(seataErrors.InsertRecordError, fmt.Sprintf("insert tcc fence record errors, prepare fence failed. xid= %s, branchId= %d", xid, branchId))
+	}
+}
+
+func (handler *tccFenceDbProxyHandler) CommitFence(ctx context.Context, tx *sql.Tx, callback func() error) error {
+	xid := tm.GetBusinessActionContext(ctx).Xid
+	branchId := tm.GetBusinessActionContext(ctx).BranchId
+
+	fenceDo := handler.tccFenceDao.QueryTCCFenceDO(tx, xid, branchId)
+	if fenceDo == nil {
+		return seataErrors.NewTccFenceError(seataErrors.RecordNotExists, fmt.Sprintf("tcc fence record not exists, commit fence method failed. xid= %s, branchId= %d ", xid, branchId))
+	}
+	if fenceDo.Status == constant.StatusCommitted {
+		log.Infof("branch transaction has already committed before. idempotency rejected. xid: %s, branchId: %d, status: %d", xid, branchId, fenceDo.Status)
+		return nil
+	}
+	if fenceDo.Status == constant.StatusRollbacked || fenceDo.Status == constant.StatusSuspended {
+		// enable warn level
+		log.Warnf("Branch transaction status is unexpected. xid: %s, branchId: %d, status: %s", xid, branchId, fenceDo.Status)
+		return nil
+	}
+	return handler.updateStatusAndInvokeTargetMethod(tx, callback, xid, branchId, constant.StatusCommitted)
+}
+
+func (handler *tccFenceDbProxyHandler) RollbackFence(ctx context.Context, tx *sql.Tx, callback func() error) error {
+	xid := tm.GetBusinessActionContext(ctx).Xid
+	branchId := tm.GetBusinessActionContext(ctx).BranchId
+	actionName := tm.GetBusinessActionContext(ctx).ActionName
+
+	fenceDo := handler.tccFenceDao.QueryTCCFenceDO(tx, xid, branchId)
+	// record is null, mean the need suspend
+	if fenceDo == nil {
+		ok := handler.InsertTCCFenceLog(tx, xid, branchId, actionName, constant.StatusSuspended)
+		log.Infof("Insert tcc fence record ok: %v. xid: %s, branchId: %d", ok, xid, branchId)
+		if ok {
+			log.Infof("TCC fence record not exists, commit fence method failed. xid= %s, branchId= %d ", xid, branchId)
+			return nil
+		} else {
+			return seataErrors.NewTccFenceError(seataErrors.InsertRecordError, fmt.Sprintf("insert tcc fence record errors, prepare fence failed. xid= %s, branchId= %d", xid, branchId))
```

Comment:
is not prepare, this step is rollbacking.

Reply:
hx, it is typo but have fixed now